relayer documentation updates (#90)
sygmaprotocol/docs: docs/03-architecture/05-security/05-keyreshare.md

---
slug: /readme/architecture/security/reshare
id: security-reshare
title: Key Resharing
description: The following details the process for key resharing.
sidebar_position: 5
draft: false
---

:::info
The following details the process for key resharing.
:::

Key resharing is an operation that makes multi-party threshold signing truly practical and secure under the realistic dynamic environment. 

This operation allows for both:
- Onboarding new relayers with new private key shares, and 
- Recalibrating threshold requirements 

This is done without the need to change the underlying private key, which is a delicate and expensive procedure.

:::warning

> Our observations in the blockchain bridging space reveal that a big portion of vulnerabilities found are exploits of the key resharing mechanism. We presume that these vulnerabilities are as a result of sensitive key material as well as the operation needing to be repeated regularly.

Reviewer: This is generally true, but IIRC vulns in key resharing can often be explained by the lack of well-defined protocols at their implementation. For ECDSA, the well-studied resharing proto was only introduced in the latest revision - [CGGMP21](https://eprint.iacr.org/2021/060).

Author: ive added the additional context!

:::

:::info
By regularly updating private key shares, the resharing mechanism can act as a defense measure from a proactive adversary. Such a key share refresh mechanism aims to prevent an attacker from consecutively compromising relayers and ultimately recovering the full private key.
:::

Similar to the initial keygen operation, resharing can also be triggered by the same governance contract onchain and propagated to the MPC relayers via emitted events. 

> However, we note that the resharing protocol requires peer discovery for both new and old relayer subsets to take place before the interactive protocol itself. Hence, onboarding relayers are required to advertise themselves to bootstrapping nodes, while the existing ones must refresh their local address books soon after that specific onchain event has been observed.

Reviewer: ```suggestion
The resharing protocol requires, before the interactive protocol, current committee members to be aware of new peers and vice versa. Given the number of security limitations of DHT-based peer discovery protocols and the overhead of running it in a small-scale network - we opted to use a static configuration that is shared securely and in advance.
```

> Also note that during this procedure, the entire set of new relayers must be active, but only a threshold number of old relayers are required to participate. The interaction flow of the resharing scheme leaves all control in the hands of old committee (relayer) participants. It is this committee that decides with whom they are gonna interact with in order to send cryptographic materials needed to eventually generate key shares. Old committee relayers must identify relayers from the new committee as prescribed by the governance contract state. For this, communication channels are required to be authenticated by the public keys that constitutes to the address stored onchain.

Reviewer: ```suggestion
Also note that during this procedure, the entire set of new relayers must be active, but only a threshold number of old relayers are required to participate. The interaction flow of the resharing scheme leaves all control in the hands of old committee (relayer) participants. This committee decides with whom they will interact to send cryptographic materials needed to generate key shares. Old committee relayers must identify relayers from the new committee as observed from the `Bridge.sol` contract. For this, communication channels are required to be authenticated by the public keys that constitute the address stored onchain.
```
